README.md
[![Build Status](https://travis-ci.org/treasure-data/embulk-input-sfdc.svg?branch=master)](https://travis-ci.org/treasure-data/embulk-input-sfdc)
[![Code Climate](https://codeclimate.com/github/treasure-data/embulk-input-sfdc/badges/gpa.svg)](https://codeclimate.com/github/treasure-data/embulk-input-sfdc)
[![Test Coverage](https://codeclimate.com/github/treasure-data/embulk-input-sfdc/badges/coverage.svg)](https://codeclimate.com/github/treasure-data/embulk-input-sfdc/coverage)
[![Gem Version](https://badge.fury.io/rb/embulk-input-sfdc.svg)](https://badge.fury.io/rb/embulk-input-sfdc)

# embulk-input-sfdc

embulk-input-sfdc is the Embulk input plugin for [Salesforce](http://www.salesforce.com/).

This plugin uses Force.com REST API.

## Overview

Required Embulk version >= 0.8.7

* **Plugin type**: input
* **Resume supported**: yes (since v0.1.1)
* **Cleanup supported**: no
* **Guess supported**: yes

## Setup

### How to obtain OAuth Client ID and Client Secret

You should create "Connected App" in SalesForce website.

First, please visit "Setup" -> "App Setup" -> "Create" -> "Apps", scroll down to "Connected Apps" and click "New" button. Please create "Connected App" its "Enable OAuth Settings" is checked and "Selected OAuth Access" is "Full access (full)". Then, this plugin doesn't use "Callback URL" setting, so please specify dummy URL (e.g. http://localhost:4000) as it.

Finally, please click you connected app name in "Connected Apps" page (visit "Setup" -> "App Setup" -> "Create" -> "Apps"). It will show "Consumer Key" (as "Client ID") and Consumer Secret (as "Client Secret").

### NOTE: security token

This plugin requires your "security token", but if you forget it, please visit "My Settings" -> "Personal" -> "Reset My Security Token", and reset it. (Please attend your API applications can't work well with old security token.)

### Configuration

- **username** username for Force.com REST API (string, required)
- **password** password for Force.com REST API (string, required)
- **client_id** client id for your application (string, required)
- **client_secret** client secret for your application (string, required)
- **security_token** your security token (string, required)
- **login_url** your login URL  (string, required)
- **target** target SObject name(string, required)
- **columns** target SObject attributes. You can generate this configuration by `guess` command (array, required)
- **retry_initial_wait_sec**: Wait seconds for exponential backoff initial value (integer, default: 1)
- **retry_limit**: Try to retry this times (integer, default: 5)
- **continue_from**: Only process records after this time. This option didn't filled or SOQL SELECT-clause doesn't contain `LastModifiedDate`, all fetched records are processed. (string, default: nil)
- **incremental**: If false, `continue_from` in next.yml would not be updated that means you always fetch all of data from SFDC with statically conditions. If true, `continue_from` would be updated in next.yml. As above you'll fetch fresh data since last executed only. (boolean, default: true)

## Example

```yaml
in:
  type: sfdc
  username: USERNAME
  password: PASSWORD
  client_id: CLIENT_ID
  client_secret: CLIENT_SECRET
  security_token: SECURITY_TOKEN
  login_url: https://your.login-url
  target: Contact
```

## Build

```
$ bundle exec rake build
```


CHANGELOG.md
## 0.2.4 - 2016-07-19
* [enhancement] Retry for some 4xx errors [#47](https://github.com/treasure-data/embulk-input-sfdc/pull/47)

## 0.2.3 - 2016-07-13
* [fixed] Fix fetched count [#46](https://github.com/treasure-data/embulk-input-sfdc/pull/46)

## 0.2.2 - 2016-07-08
* [enhancement] retrieve child-to-parent relationship columns [#45](https://github.com/treasure-data/embulk-input-sfdc/pull/45) thanks @noissefnoc

## 0.2.1 - 2016-07-06
* [enhancement] Display log for each page fetched to avoid silence on long running job [#44](https://github.com/treasure-data/embulk-input-sfdc/pull/44)

## 0.2.0 - 2016-07-05
**This version drops old Embulk versions. Embulk version should be >= 0.8.7**

* [enhancement] Memory efficient processing [#43](https://github.com/treasure-data/embulk-input-sfdc/pull/43)

## 0.1.4 - 2016-03-07
* [fixed] Check searchable is not needed to this plugin [#42](https://github.com/treasure-data/embulk-input-sfdc/pull/42)

## 0.1.3 - 2016-01-18
* [fixed] Fix `incremental` option when specified in config.yml


## 0.1.2 - 2016-01-13
* [maintenance] relax httpclient version [#40](https://github.com/treasure-data/embulk-input-sfdc/pull/40)
* [enhancement] Add `incremental` option [#41](https://github.com/treasure-data/embulk-input-sfdc/pull/41)

## 0.1.1 - 2015-11-16

* [enhancement] Scheduled execution [#39](https://github.com/treasure-data/embulk-input-sfdc/pull/39)
* [maintenance] Set retryer logger as Embulk.logger [#38](https://github.com/treasure-data/embulk-input-sfdc/pull/38)

## 0.1.0 - 2015-11-09

* [enhancement] Add retry options and retry handling [#37](https://github.com/treasure-data/embulk-input-sfdc/pull/37)

## 0.0.7 - 2015-10-21

* [enhancement] Better http error detection [#36](https://github.com/treasure-data/embulk-input-sfdc/pull/36)
* [maintenance] Rename: commit_report -> task_report renamed by Embulk since 0.7 [#34](https://github.com/treasure-data/embulk-input-sfdc/pull/34)

## 0.0.6 - 2015-10-06

* [enhancement] Supports embulk.0.7 [#33](https://github.com/treasure-data/embulk-input-sfdc/pull/33)
* [maintenance] updated description of gemspec to be more descriptive [#32](https://github.com/treasure-data/embulk-input-sfdc/pull/32)

## 0.0.5 - 2015-08-24

* [enhancement] Convert timestamp value when preivew/run [#30](https://github.com/treasure-data/embulk-input-sfdc/pull/30)

## 0.0.4 - 2015-08-18

**Embulk 0.6.13+ is required since this version**

* [enhancement] raise ConfigError if config.yml is wrong (retry is unneeded) [#27](https://github.com/treasure-data/embulk-input-sfdc/pull/27)
* [enhancement] Improve config.param validations by ConfigError [#25](https://github.com/treasure-data/embulk-input-sfdc/pull/25)
* [maintenance] Brush up some code [#24](https://github.com/treasure-data/embulk-input-sfdc/pull/24) [#26](https://github.com/treasure-data/embulk-input-sfdc/pull/26)
* [maintenance] Add setup to readme [#22](https://github.com/treasure-data/embulk-input-sfdc/pull/22)
* [maintenance] Put all dependency gems into gemspec [#21](https://github.com/treasure-data/embulk-input-sfdc/pull/21)

## 0.0.3 - 2015-06-24

**Embulk 0.6.12+ is required since this version**

* [enhancement] Follow embulk 0.6.12 [#19](https://github.com/treasure-data/embulk-input-sfdc/pull/19)

## 0.0.2 - 2015-06-22
* [enhancement] Limitation for guess and preview [#17](https://github.com/treasure-data/embulk-input-sfdc/pull/17)
* [enhancement] Show messages while `run` command executing [#14](https://github.com/treasure-data/embulk-input-sfdc/pull/14)

## 0.0.1 - 2015-06-16

The first release!!


test/run-test.rb
#!/usr/bin/env ruby

base_dir = File.expand_path(File.join(File.dirname(__FILE__), ".."))
lib_dir = File.join(base_dir, "lib")
test_dir = File.join(base_dir, "test")

require "test-unit"
require "test/unit/rr"

$LOAD_PATH.unshift(lib_dir)
$LOAD_PATH.unshift(test_dir)

ENV["TEST_UNIT_MAX_DIFF_TARGET_STRING_SIZE"] ||= "5000"

exit Test::Unit::AutoRunner.run(true, test_dir, ARGV + ["--use-color=yes"])


test/sfdc/test-api.rb
require "sfdc/api"

class SfdcApiTest < Test::Unit::TestCase
  def test_setup
    config = {}

    any_instance_of(Sfdc::Api) do |klass|
      mock(klass).authentication(config) { "access_token" }
      mock(klass).set_latest_version("access_token") { klass }
    end

    Sfdc::Api.setup("login_url", config)
  end

  def test_initialize
    api = Sfdc::Api.new("login_url")

    assert_true(api.client.is_a?(HTTPClient))
  end
end


test/output_capture.rb
module OutputCapture
  def capture(output = :out, &block)
    _, out = swap_io(output, &block)
    out
  end

  def silence(&block)
    block_result = nil
    swap_io(:out) do
      block_result,_ = swap_io(:err, &block)
    end
    block_result
  end

  def swap_io(output = :out, &block)
    java_import 'java.io.PrintStream'
    java_import 'java.io.ByteArrayOutputStream'
    java_import 'java.lang.System'

    ruby_original_stream = output == :out ? $stdout.dup : $stderr.dup
    java_original_stream = System.send(output) # :out or :err
    ruby_buf = StringIO.new
    java_buf = ByteArrayOutputStream.new

    case output
    when :out
      $stdout = ruby_buf
      System.setOut(PrintStream.new(java_buf))
    when :err
      $stderr = ruby_buf
      System.setErr(PrintStream.new(java_buf))
    end

    [block.call, ruby_buf.string + java_buf.toString]
  ensure
    case output
    when :out
      $stdout = ruby_original_stream
      System.setOut(java_original_stream)
    when :err
      $stderr = ruby_original_stream
      System.setErr(java_original_stream)
    end
  end
end


lib/embulk/input/sfdc_api/api.rb
require "httpclient"
require "pathname"

module Embulk
  module Input
    module SfdcApi
      class InternalServerError < StandardError; end

      class Api
        DEFAULT_LOGIN_URL = "https://login.salesforce.com".freeze

        attr_reader :client

        def setup(login_url, config)
          token = authentication(login_url, config)
          set_latest_version(token)
          self
        end

        def initialize
          @version_path = ""
          @client = HTTPClient.new
          @client.default_header = {Accept: 'application/json; charset=UTF-8'}
        end

        def get(path, parameters={})
          response = catch_unretryable_error do
            client.get(path, parameters)
          end
          handle_error(response)
          JSON.parse(response.body)
        end

        def get_metadata(sobject_name)
          get(@version_path.join("sobjects/#{sobject_name}/describe").to_s)
        end

        def search(soql)
          get(@version_path.join("query").to_s, {q: soql})
        end

        private

        def catch_unretryable_error(&block)
          # if can't resolve a problem with retry, should raise ConfigError to tell Embulk
          begin
            yield
          rescue SocketError => e # probably login_url is wrong
            raise ConfigError.new "SocketError: #{e.message}"
          end
        end

        def authentication(login_url, _config)
          # NOTE: At SfdcInputPlugin#init, we use Symbol as each key
          #       for task (Hash), but at SfdcInputPlugin#run, task
          #       has them as String...:(
          #       So, I translate keys from String to Symbol
          config = {}
          _config.each { |key, value| config[key.to_sym] = value }

          params = {
            grant_type: 'password',
            client_id: config[:client_id],
            client_secret: config[:client_secret],
            username: config[:username],
            password: config[:password] + config[:security_token]
          }

          oauth_response = catch_unretryable_error do
            @client.post(URI.join(login_url, "services/oauth2/token").to_s, params)
          end
          handle_error(oauth_response)
          oauth = JSON.parse(oauth_response.body)

          client.base_url = oauth["instance_url"]

          oauth["access_token"]
        end

        def set_latest_version(access_token)
          versions_response = @client.get("/services/data")
          # Use latest version always
          @version_path = Pathname.new(JSON.parse(versions_response.body).last["url"])

          client.default_header = client.default_header.merge(Authorization: "Bearer #{access_token}")

          self
        end

        def handle_error(response)
          code = response.status_code

          case code
          when 400..499
            message = "StatusCode: #{code}"

            body = nil
            begin
              body = JSON.parse(response.body)
            rescue
              # Sometimes SFDC returns non-json response
              # https://github.com/treasure-data/embulk-input-sfdc/issues/35
              message << ": #{response.body}"
              raise ConfigError.new message
            end

            message << ": #{body['errorCode']}" if body["errorCode"]
            message << ": #{body['message']}" if body["message"]
            case body["errorCode"]
            when "INVALID_QUERY_LOCATOR", "QUERY_TIMEOUT"
              # will be retried
              raise message
            else
              # won't retry
              raise ConfigError.new message
            end
          when 500..599
            raise SfdcApi::InternalServerError, "Force.com REST API returns 500 (Internal Server Error). Please contact customer support of Force.com."
          end
        end
      end
    end
  end
end


lib/embulk/input/sfdc_input_plugin.rb
require "embulk/input/sfdc/api"
require "embulk/input/sfdc_input_plugin_utils"
require "logger"

module Embulk
  module Input

    class SfdcInputPlugin < InputPlugin
      Plugin.register_input("sfdc", self)

      GUESS_RECORDS_COUNT = 30
      PREVIEW_RECORDS_COUNT = 15
      MAX_FETCHABLE_COUNT = 2000

      def self.transaction(config, &control)
        task = {}

        task[:login_url] = config.param("login_url", :string, default: Sfdc::Api::DEFAULT_LOGIN_URL)
        task[:soql] = config.param("soql", :string)

        task[:config] = embulk_config_to_hash(config)

        task[:schema] = config.param("columns", :array)
        columns = []

        task[:schema].each do |column|
          name = column["name"]
          type = column["type"].to_sym

          columns << Column.new(nil, name, type, column["format"])
        end

        resume(task, columns, 1, &control)
      end

      def self.resume(task, columns, count, &control)
        commit_reports = yield(task, columns, count)

        next_config_diff = {}
        return next_config_diff
      end

      def self.guess(config)
        login_url = config.param("login_url", :string, default: Sfdc::Api::DEFAULT_LOGIN_URL)
        target = config.param("target", :string)

        api = Sfdc::Api.new.setup(login_url, embulk_config_to_hash(config))

        metadata = api.get_metadata(target)
        raise "Target #{target} can't be searched." unless searchable_target?(metadata)

        soql = SfdcInputPluginUtils.build_soql(target, metadata)
        sobjects = api.search("#{soql} LIMIT #{GUESS_RECORDS_COUNT}")

        {
          "soql" => soql,
          "columns" => SfdcInputPluginUtils.guess_columns(sobjects)
        }
      end

      def init
        @api = Sfdc::Api.new.setup(task["login_url"], task["config"])
        @schema = task["schema"]
        @soql = task["soql"]
      end

      def run
        @soql += " LIMIT #{PREVIEW_RECORDS_COUNT}" if preview?
        response = @api.search(@soql)
        logger.debug "Start to add records...(total #{response["totalSize"]} records)"
        add_records(response["records"])

        add_next_records(response, 1)

        page_builder.finish

        logger.debug "Added all records."

        commit_report = {}
        return commit_report
      end

      private

      def preview?
        # NOTE: This is workaround for "org.embulk.spi.Exec.isPreview"
        # TODO: Extract process for preview command to method
        begin
          org.embulk.spi.Exec.session().isPreview()
        rescue java.lang.NullPointerException => e
          false
        end
      end

      def self.embulk_config_to_hash(config)
        {
          client_id: config.param("client_id", :string),
          client_secret: config.param("client_secret", :string),
          username: config.param("username", :string),
          password: config.param("password", :string),
          security_token: config.param("security_token", :string),
        }
      end

      def self.searchable_target?(metadata)
        !!(metadata["queryable"] && metadata["searchable"])
      end

      def add_next_records(response, fetch_count)
        return if response["done"]
        logger.debug "Added #{MAX_FETCHABLE_COUNT * fetch_count}/#{response["totalSize"]} records."
        next_url = response["nextRecordsUrl"]
        response = @api.get(next_url)

        add_records(response["records"])

        add_next_records(response, fetch_count + 1)
      end

      def add_records(records)
        records = SfdcInputPluginUtils.extract_records(records)

        records.each do |record|
          values = @schema.collect do |column|
            SfdcInputPluginUtils.cast(record[column["name"]], column["type"])
          end

          page_builder.add(values)
        end
      end

      # TODO: Replace Embulk::Logger after merging
      # https://github.com/embulk/embulk/pull/200
      def self.logger
        Embulk.logger
      end

      def logger
        self.class.logger
      end
    end
  end
end


lib/embulk/input/sfdc-input-plugin-utils.rb
# This module contains methods for Plugin.

module Embulk
  module Input
    module SfdcInputPluginUtils
      # Guess::SchemaGuess.from_hash_records returns Columns
      # containing 'index' key, but it is needless.
      def self.guess_columns(records)
        schema = Guess::SchemaGuess.from_hash_records(records)

        schema.map do |c|
          column = {name: c.name, type: c.type}
          column[:format] = c.format if c.format
          column
        end
      end

      # NOTE: Force.com query API returns JSON including
      #       sobject name (a.k.a "attributes") and record data.
      def self.extract_records(json)
        json.map do |elements|
          elements.reject {|key, _| key == "attributes" }
        end
      end

      def self.cast(value, type)
        return value if value.nil?

        case type.to_sym
        when :long
          Integer(value)
        when :double
          Float(value)
        when :timestamp
          Time.parse(value)
        when :boolean
          !!value
        else
          value.to_s
        end
      end
    end
  end
end


lib/embulk/input/sfdc_input_plugin_utils.rb
# This module contains methods for Plugin.

module Embulk
  module Input
    module SfdcInputPluginUtils
      # Guess::SchemaGuess.from_hash_records returns Columns
      # containing 'index' key, but it is needless.
      def self.guess_columns(sobjects)
        records = extract_records(sobjects["records"])
        schema = Guess::SchemaGuess.from_hash_records(records)

        schema.map do |c|
          column = {name: c.name, type: c.type}
          column[:format] = c.format if c.format
          column
        end
      end

      def self.build_soql(target, metadata)
        target_columns = metadata["fields"].map {|fields| fields["name"] }
        "SELECT #{target_columns.join(',')} FROM #{target}"
      end

      # NOTE: Force.com query API returns JSON including
      #       sobject name (a.k.a "attributes") and record data.
      #       record data has 3 types
      #         1. self columns
      #         2. parent's columns (follow child-to-parent relationship)
      #         3. child's columns (follow parent-to-child relationship)
      def self.extract_records(json, schema={})
        type_json_columns = schema_to_type_json_columns(schema)

        json.map do |elements|
          record = {}

          elements.each {|key, value|
            if key != "attributes"
              if value.is_a?(Hash) and !schema.empty?
                # for parent's columns (expand json)
                if value.has_key?("attributes") and !type_json_columns.has_key?(key)
                  record.merge!(extract_parent_elements(key, value, type_json_columns))
                # for child's columns, parent's columns (as json) or something else
                else
                  record[key] = value
                end
              # for self columns
              else
                record[key] = value
              end
            end
          }
          record
        end
      end

      def self.extract_parent_elements(key, elements, type_json_columns)
        record = {}

        elements.each{|k, v|
          full_key_name = key + '.' + k

          if k != "attributes"
            if v.is_a?(Hash)
              # for parent's columns (expand json)
              # follow child-to-parent relationship recursivly
              if v.has_key?("attributes") and !type_json_columns.has_key?(full_key_name)
                record.merge!(extract_parent_elements(full_key_name, v, type_json_columns))
              # for parent's columns (as json)
              else
                record[full_key_name] = v
              end
            # for self columns
            else
              record[full_key_name] = v
            end
          end
        }
        record
      end

      def self.schema_to_type_json_columns(schema)
        columns = {}
        schema.map{|column| columns[column["name"]] = 1 if column["type"] == "json" }
        columns
      end
    end
  end
end


lib/embulk/input/sfdc.rb
require "perfect_retry"
require "embulk/input/sfdc_api/api"
require "embulk/input/sfdc_input_plugin_utils"

module Embulk
  module Input

    class Sfdc < InputPlugin
      Plugin.register_input("sfdc", self)

      GUESS_RECORDS_COUNT = 30
      PREVIEW_RECORDS_COUNT = 15

      def self.transaction(config, &control)
        task = {}

        task[:login_url] = config.param("login_url", :string, default: SfdcApi::Api::DEFAULT_LOGIN_URL)
        task[:soql] = config.param("soql", :string)

        task[:config] = embulk_config_to_hash(config)
        task[:retry_limit] = config.param("retry_limit", :integer, default: 5)
        task[:retry_initial_wait_sec] = config.param("retry_initial_wait_sec", :integer, default: 1)
        task[:continue_from] = config.param("continue_from", :string, default: nil)

        task[:schema] = config.param("columns", :array)
        task[:incremental] = config.param("incremental", :bool, default: true)
        columns = []

        task[:schema].each do |column|
          name = column["name"]
          type = column["type"].to_sym

          columns << Column.new(nil, name, type, column["format"])
        end

        resume(task, columns, 1, &control)
      end

      def self.resume(task, columns, count, &control)
        task_reports = yield(task, columns, count)

        return task_reports.first
      end

      def self.guess(config)
        login_url = config.param("login_url", :string, default: SfdcApi::Api::DEFAULT_LOGIN_URL)
        target = config.param("target", :string)

        api = SfdcApi::Api.new.setup(login_url, embulk_config_to_hash(config))

        metadata = api.get_metadata(target)
        raise "Target #{target} can't be searched." unless searchable_target?(metadata)

        soql = SfdcInputPluginUtils.build_soql(target, metadata)
        sobjects = api.search("#{soql} LIMIT #{GUESS_RECORDS_COUNT}")

        {
          "soql" => soql,
          "columns" => SfdcInputPluginUtils.guess_columns(sobjects)
        }
      end

      def init
        @api = SfdcApi::Api.new.setup(task["login_url"], task["config"])
        @schema = task["schema"]
        @soql = task["soql"]
        @retryer = PerfectRetry.new do |config|
          config.limit = task[:retry_limit]
          config.sleep = proc{|n| task[:retry_initial_wait_sec] ** n}
          config.dont_rescues = [Embulk::ConfigError]
          config.logger = Embulk.logger
          config.log_level = nil
        end
      end

      def run
        @soql += " LIMIT #{PREVIEW_RECORDS_COUNT}" if preview?
        if task[:continue_from]
          @continue_from = @latest_updated = Time.parse(task[:continue_from])
        end

        response = @retryer.with_retry do
          @api.search(@soql)
        end
        Embulk.logger.info "Start to add records...(total #{response["totalSize"]} records)"
        add_records(response["records"])

        add_next_records(response, 1) unless preview?

        page_builder.finish

        Embulk.logger.debug "Added all records."

        return {} unless task[:incremental]

        task_report = {
          continue_from: @latest_updated.to_s
        }
        return task_report
      end

      private

      def preview?
        begin
          org.embulk.spi.Exec.isPreview()
        rescue java.lang.NullPointerException => e
          false
        end
      end

      def self.embulk_config_to_hash(config)
        {
          client_id: config.param("client_id", :string),
          client_secret: config.param("client_secret", :string),
          username: config.param("username", :string),
          password: config.param("password", :string),
          security_token: config.param("security_token", :string),
        }
      end

      def self.searchable_target?(metadata)
        !!metadata["queryable"]
      end

      def add_next_records(response, fetch_count)
        fetched_count = 0
        loop do
          fetched_count += response["records"].length
          Embulk.logger.info "Fetched #{fetched_count}/#{response["totalSize"]} records."
          break if response["done"]
          next_url = response["nextRecordsUrl"]

          response = @retryer.with_retry do
            @api.get(next_url)
          end

          add_records(response["records"])
          fetch_count += 1
        end
      end

      def add_records(records)
        records = SfdcInputPluginUtils.extract_records(records, @schema)

        records.each do |record|
          if record.has_key?("LastModifiedDate")
            updated_at = Time.parse(record["LastModifiedDate"])
            set_latest_updated_at(updated_at)

            if @continue_from && @continue_from >= updated_at
              Embulk.logger.warn "'#{updated_at}'(LastModifiedDate) is earlier than or equal to '#{@continue_from}'(continue_from). Skipped"
              next
            end
          end

          values = record_to_values(record)
          page_builder.add(values)
        end
      end

      def set_latest_updated_at(updated_at)
        @latest_updated = [
          @latest_updated || updated_at,
          updated_at
        ].max
      end

      def record_to_values(record)
        @schema.collect do |column|
          val = record[column["name"]]
          if column["type"] == "timestamp" && val
            begin
              val = Time.parse(val.to_s)
            rescue ArgumentError => e # invalid date
              raise ConfigError.new "The value '#{val}' (as '#{column['name']}') is invalid time format"
            end
          elsif val.is_a?(Hash)
            val = val.to_s
          end
          val
        end
      end
    end
  end
end
